docs/src/constraints.md

```diff
@@ -1,5 +1,261 @@
+[Variable containers](@ref), JuMP provides a mechanism for building groups of
+constraints in *containers*. Three types of constraint containers are supported:
+`Array`s, `JuMPArray`s, and `Dict`ionaries. We explain each of these in the
+following.
+
+### [Arrays](@id constraint_arrays)
+
+The syntax for adding `Array`s of JuMP constraints is very similar to the
+[syntax for creating `Array`s of JuMP variables](@ref Arrays):
+```jldoctest constraint_arrays; setup=:(model=Model(); @variable(model, x))
+julia> @constraint(model, con[i = 1:3], i * x <= i + 1)
+3-element Array{ConstraintRef{Model,C,Shape} where Shape<:JuMP.AbstractShape where C,1}:
+ con[1] : x <= 2.0
+ con[2] : 2 x <= 3.0
+ con[3] : 3 x <= 4.0
+```
+These arrays can be accessed and sliced like normal Julia arrays:
+```jldoctest constraint_arrays
+julia> con[1]
+con[1] : x <= 2.0
+
+julia> con[2:3]
+2-element Array{ConstraintRef{Model,C,Shape} where Shape<:JuMP.AbstractShape where C,1}:
+ con[2] : 2 x <= 3.0
+ con[3] : 3 x <= 4.0
+```
+
+Anonymous containers can also be constructed by dropping the name (e.g. `con`)
+before the square brackets:
+```jldoctest constraint_arrays
+julia> @constraint(model, [i = 1:2], i * x <= i + 1)
+2-element Array{ConstraintRef{Model,C,Shape} where Shape<:JuMP.AbstractShape where C,1}:
+ x <= 2.0
+ 2 x <= 3.0
+```
+
+Just like [`@variable`](@ref), JuMP will form an `Array` of constraints when it
+can determine at compile time that the indices are one-based integer ranges.
+Therefore `con[1:b]` will work, but `con[a:b]` will throw an error. If JuMP
+cannot determine that the indices are one-based integer ranges, JuMP will
+constraint a `JuMPArray` instead.
+
+### JuMPArrays
+
+The syntax for constructing a `JuMPArray` of constraints is very similar to the
+[syntax for constructing](@ref variable_jump_arrays) a JuMPArray of variables.
+
+```jldoctest constraint_jumparrays; setup=:(model=Model(); @variable(model, x))
+julia> @constraint(model, con[i = 1:2, j = 2:3], i * x <= j + 1)
+2-dimensional JuMPArray{ConstraintRef{Model,C,Shape} where Shape<:JuMP.AbstractShape where C,2,...} with index sets:
+    Dimension 1, 1:2
+    Dimension 2, 2:3
+And data, a 2×2 Array{ConstraintRef{Model,C,Shape} where Shape<:JuMP.AbstractShape where C,2}:
+ con[1,2] : x <= 3.0    con[1,3] : x <= 4.0
+ con[2,2] : 2 x <= 3.0  con[2,3] : 2 x <= 4.0
+```
+
+### Dictionaries
+
+The syntax for constructing a dictionary of constraints is very similar to the
+[syntax for constructing](@ref variable_dictionaries) a dictionary of variables.
+
+```jldoctest constraint_jumparrays; setup=:(model=Model(); @variable(model, x))
+julia> @constraint(model, con[i = 1:2, j = 1:2; i != j], i * x <= j + 1)
+Dict{Any,ConstraintRef{Model,C,Shape} where Shape<:JuMP.AbstractShape where C} with 2 entries:
+  (1, 2) => con[1,2] : x <= 3.0
+  (2, 1) => con[2,1] : 2 x <= 2.0
+```
+
+### Forcing the container type
+
+When creating a container of constraints, JuMP will attempt to choose the
```

Comment:
We should only have this discussion once. Can we just say that it's analogous to http://www.juliaopt.org/JuMP.jl/dev/variables/#Forcing-the-container-type-1?